### ISSUE TITLE
App is lagging when initializing Google Maps

### ISSUE THREAD

**Issue by ItsPierre** [flutter/flutter#28493]
## Steps to Reproduce
I am just building the Google-Maps widget in my Stateful Widget. This will be opened with the Navigator via a Button click. But as soon, as I click, the app is lagging, while initializing the map.
When removing the widget, everything is fine.

I tried to set a CircularProgressIndicator and then, setting the map-widget async, but running into the same problem, as soon, as it gets build.
I also downloaded and ran the example app from the google-maps-package with the same lagging.

How should the map be initilized, that it is first loaded and then set without lagging?

**Comment by zoechi** [comment]
Please add the output of `flutter doctor -v`.

A minimal reproduction would be helpful as well.

**Comment by zoechi** [comment]
Have you tried with a release build? Debug builds are never representative for measuring performance.

**Comment by ItsPierre** [comment]
> Please add the output of `flutter doctor -v`.

I will add the `flutter doctor -v` later

> Have you tried with a release build? Debug builds are never representative for measuring performance.

Yes, I did try release builds, but it is still lagging, when initializing

**Comment by ItsPierre** [comment]
> Please add the output of `flutter doctor -v`.

I added the flutter doctor. But I think, this is a general problem.

**Comment by zoechi** [comment]
> But I think, this is a general problem.

It's not about possible errors `flutter doctor` might report. 
`flutter doctor -v` output is valuable information about the context where the problem reproduced for you.

**Comment by realvotum** [comment]
Same for me. It takes a while to build my StatelessWidget with google map inside SizedBox 300x300. Without google maps, the widget builds instantly, otherwise the delay for the StatelessWidget to appear is noticeable. This is my output:

[✓] Flutter (Channel stable, v1.2.1, on Mac OS X 10.14.3 18D109, locale pl-PL)
    • Flutter version 1.2.1 at /Users/mat/dev/flutter
    • Framework revision 8661d8aecd (5 weeks ago), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/mat/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 33.3.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.1.0 (API 27) (emulator)

• No issues found!

**Comment by JRR-OSU** [comment]
Experiencing the same in my release build.

**Comment by eopeter** [comment]
For now I added a delay in initState to load Google Maps:

Future.delayed(const Duration(milliseconds: 500), () {
      setState(() {
        _showGoogleMaps = true;
      });
    });

**Comment by danielcmm** [comment]
Using this plugin really degrades the performance of the Widget tree for me... i had to put the map on a separate, isolate screen ;/

**Comment by dshukertjr** [comment]
@danielcmm How do you put a widget on a separate isolate?

**Comment by devmgs** [comment]
> For now I added a delay in initState to load Google Maps:
> 
> Future.delayed(const Duration(milliseconds: 500), () {
> setState(() {
> _showGoogleMaps = true;
> });
> });

Works nice. Good Workaround utill it is fixed.

**Comment by Hixie** [comment]
It would be good to study a timeline showing what is actually happening. Can someone who is seeing this run a profile build and collect a timeline of the behaviour?

It would also be good to see a short app that reproduces the problem, that would save us some time when we get around to debugging this.

**Comment by deckerst** [comment]
I tried to make a simple app to show the issue:
```dart
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final controller = PageController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          children: [
            Expanded(
              child: PageView(
                controller: controller,
                children: [
                  Container(
                    decoration: BoxDecoration(
                      gradient: LinearGradient(stops: [0, 1], colors: [Colors.blue, Colors.transparent]),
                    ),
                  ),
                  MapPage(),
                ],
              ),
            ),
            RaisedButton(
              onPressed: () => controller.animateToPage(
                controller.page == 0 ? 1 : 0,
                duration: Duration(milliseconds: 350),
                curve: Curves.linear,
              ),
              child: Text('Switch page'),
            )
          ],
        ),
      ),
    );
  }
}

class MapPage extends StatefulWidget {
  @override
  _MapPageState createState() => _MapPageState();
}

class _MapPageState extends State<MapPage> with AutomaticKeepAliveClientMixin {
  bool _keepAlive = false;

  @override
  Widget build(BuildContext context) {
    super.build(context);
    return SafeArea(
      child: ListView(
        children: [
          Row(
            children: [
              Checkbox(
                value: _keepAlive,
                onChanged: (v) {
                  _keepAlive = v;
                  updateKeepAlive();
                  setState(() {});
                },
              ),
              Text('Keep alive'),
            ],
          ),
          SizedBox(
            height: 300,
            child: GoogleMap(
              initialCameraPosition: CameraPosition(
                target: LatLng(37.42796133580664, -122.085749655962),
              ),
            ),
          ),
        ],
      ),
    );
  }

  @override
  bool get wantKeepAlive => _keepAlive;
}
```

The app shows a `PageView` that can switch to a page with a map. It highlights 3 different cases:
1. first initialization: the very first time it goes to the map page, the app lags
2. following initializations without keep alive: the map is reinitialized every time but it does not lag (the map is grey for a few frames during initialization, which is normal)
3. following initializations with keep alive: actually the map is not reinitialized, so it's always ready and smooth.

Here's a timeline of the first initialization with lag:
[timeline_2020_1_18-1579319582040000.json.txt](https://github.com/flutter/flutter/files/4079947/timeline_2020_1_18-1579319582040000.json.txt)

Please note that it's the first time I check and export a timeline, so let me know if it's not what you expected...

**Comment by kras95** [comment]
> > For now I added a delay in initState to load Google Maps:
> > Future.delayed(const Duration(milliseconds: 500), () {
> > setState(() {
> > _showGoogleMaps = true;
> > });
> > });
> 
> Works nice. Good Workaround utill it is fixed.

This didn't help for me, but not sure I'm doing it right.

I have a GoogleMap in a ListView.

If _showGoogleMaps is true, I return the GoogleMap() in the ListView, otherwise I return an empty Container(), when the 500 millisecond timeout is done, the map still lags when initiating.

Does anybody have other solutions? Is there any way to async load the map before it's actually displayed, or maybe load the map before actually reaching the map in the ListView(), because the moment the ListView() scrolls to the GoogleMap(), the app just freezes until the GoogleMap() is rendered. If the map renders once, scrolling back to it doesn't cause lags.

**Comment by otopba** [comment]
This can help 
https://github.com/flutter/plugins/pull/2479

**Comment by shlima** [comment]
The same issue

**Comment by ndhbr** [comment]
Any updates?

**Comment by shlima** [comment]
@dshukertjr  I've made map initialization inside isolate:

```dart
import 'dart:async';
import 'package:meta/meta.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

class GoogleMapInitializer extends StatelessWidget {
    GoogleMapInitializer({
        @required this.child
    });

    final Widget child;

    @override
    build(BuildContext context) {
        return StreamBuilder(
            stream: Stream.fromFuture(compute(_preload, null)),
            initialData: false,
            builder: (context, snapshot) {
                if (snapshot.hasError) {
                    throw snapshot.error;
                } else if (snapshot.connectionState == ConnectionState.done) {
                    return child;
                } else {
                    return Center(child: LoadingIndicator(size: 40.0));
                }
            }
        );
    }
}

Future<void> _preload(_) async {
    final Completer<GoogleMapController> _controller = Completer();

    GoogleMap(
        compassEnabled: false,
        mapToolbarEnabled: false,
        rotateGesturesEnabled: false,
        scrollGesturesEnabled: false,
        zoomGesturesEnabled: false,
        tiltGesturesEnabled: false,
        buildingsEnabled: false,
        initialCameraPosition: CameraPosition(target: LatLng(0, 0), zoom: 20),
        onMapCreated: (GoogleMapController controller) {
            _controller.complete(controller);
        }
    );

    await _controller.future;
    return Future<void>.value();
}
```

Usage:

```dart
GoogleMapInitializer(
   child: GoogleMap()
)
```

**Comment by cohenadair** [comment]
> For now I added a delay in initState to load Google Maps:
> 
> Future.delayed(const Duration(milliseconds: 500), () {
> setState(() {
> _showGoogleMaps = true;
> });
> });

Thanks for the suggestion. This is working for now. I used a `FutureBuilder` instead, though. For those who are curious:

```
  final Completer<GoogleMapController> _mapController = Completer();
  Future _mapFuture = Future.delayed(Duration(milliseconds: 250), () => true);

  Widget _buildMap() {
    LatLng currentLocation = LocationMonitor.of(context).currentLocation;

    return FutureBuilder(
      future: _mapFuture,
      builder: (context, snapshot) {
        if (!snapshot.hasData) {
          print("empty");
          return Empty();
        }

        return GoogleMap(
          initialCameraPosition: CameraPosition(
            target: currentLocation == null ? LatLng(0.0, 0.0) : currentLocation,
            zoom: currentLocation == null ? 0 : 15,
          ),
          onMapCreated: (GoogleMapController controller) {
            _mapController.complete(controller);
          },
          myLocationButtonEnabled: false,
          myLocationEnabled: true,
        );
      },
    );
  }
```

**Comment by kf6gpe** [comment]
`FutureBuilder` looks like a reasonable work-around, given the time it takes the map view under the plugin to initialize.  I'm going to move this out to Stretch Goals at this point, as we don't currently have resources to work on optimizations here.

**Comment by codepushr** [comment]
@cohenadair So basically with the Future builder you are postponing the map draw until the Navigator animation is completed?

**Comment by cohenadair** [comment]
> @cohenadair So basically with the Future builder you are postponing the map draw until the Navigator animation is completed?

That's the idea. Technically, it's not waiting for the Navigator to finish, it's just delaying the map draw 250 ms. As long as the delay is longer than the Navigator animation, everything should be fine.

You may be able to listen for the Navigator animation to finish, so the delay becomes irrelevant, but I didn't look into it, and the 250 ms delay works fine for what I'm trying to do.

**Comment by DjangoOverloard** [comment]
But even when the map is initialized, I still have a lot of lag when moving through the map. This experience is fine in native google maps app on any device, but even if I try to release build my app from flutter, any android device is showing a lot of lag. Not an issue on IOS though.

**Comment by Xgamefactory** [comment]
> > @cohenadair So basically with the Future builder you are postponing the map draw until the Navigator animation is completed?
> 
> @codepushr That's the idea. Technically, it's not waiting for the Navigator to finish, it's just delaying the map draw 250 ms. As long as the delay is longer than the Navigator animation, everything should be fine.
> 
> You may be able to listen for the Navigator animation to finish, so the delay becomes irrelevant, but I didn't look into it, and the 250 ms delay works fine for what I'm trying to do.

yes exactly it would be nice to able to handle route navigation finish like initState methods for example onRoutingFinished and after that initliazing google maps

**Comment by xnio94** [comment]
hope this also helps: 

 ```
 bool showMap = false;
 void initState() {
    super.initState();
    Future.delayed(const Duration(milliseconds: 250), () {
      setState(() {
        showMap= true;
      });
    });
  }
```
and in the build use : 

  ```
 Stack(
   children:[
       if (showMap)  GoogleMap,
       Cover,
    ]
)
  ```
where GoogleMap is the GoogleMap Widget and Cover is something like this :

  ```
            IgnorePointer(
                  child: AnimatedOpacity(
                    curve: Curves.easeInCirc,
                    duration: Duration(milliseconds: 1000),
                    opacity: showMap ? 0 : 1,
                    child: Container(
                      color: Colors.blue,
                      child: Center(
                        child: Icon(
                          FontAwesomeIcons.spinner,
                          size: 60,
                          color: Colors.white.withAlpha(100),
                        ),
                      ),
                    ),
                  ),
                ),
```
this will add about 250ms before google map start initializing (this will smooth the navigator animation as many suggested) 
then it adds a one second transition to smooth the appearance of the map
it worked well for my use case

**Comment by 0ttik** [comment]
> Thanks for the suggestion. This is working for now. I used a `FutureBuilder` instead, though. For those who are curious:
> 
> ```
>   final Completer<GoogleMapController> _mapController = Completer();
>   Future _mapFuture = Future.delayed(Duration(milliseconds: 250), () => true);
> 
>   Widget _buildMap() {
>     LatLng currentLocation = LocationMonitor.of(context).currentLocation;
> 
>     return FutureBuilder(
>       future: _mapFuture,
>       builder: (context, snapshot) {
>         if (!snapshot.hasData) {
>           print("empty");
>           return Empty();
>         }
> 
>         return GoogleMap(
>           initialCameraPosition: CameraPosition(
>             target: currentLocation == null ? LatLng(0.0, 0.0) : currentLocation,
>             zoom: currentLocation == null ? 0 : 15,
>           ),
>           onMapCreated: (GoogleMapController controller) {
>             _mapController.complete(controller);
>           },
>           myLocationButtonEnabled: false,
>           myLocationEnabled: true,
>         );
>       },
>     );
>   }
> ```

I think more cleaner solution is not to w8 some random 250ms but to resolve future at the post frame callback. 

So its gonna go like this: 
```
class GodState extends State<God> {
  Completer<void> _initCompleter = Completer();

  @override
  onInitState() {
    WidgetsBinding.instance
        .addPostFrameCallback((_) => _initCompleter.complete());
  }
  
  FutureBuilder(
      future: _initCompleter.future,
      builder: (params) => BuildMe()
  );
}
```

**Comment by cohenadair** [comment]
> Related message
> I think more clean solution is not to w8 random 250ms but to resolve future at the post frame callback.
> 
> Smth like this:
> 
> ```
> class GodState extends State<God> {
>   Completer<void> _initCompleter = Completer();
> 
>   @override
>   onInitState() {
>     WidgetsBinding.instance
>         .addPostFrameCallback((_) => _initCompleter.complete());
>   }
>   
>   FutureBuilder(
>       future: _initCompleter.future,
>       builder: (...) => ...
>   );
> }
> ```

Hi @0ttik,

Can you explain how this works a little more? The documentation states that 
> This callback is run during a frame, just after the persistent frame callbacks (which is when the main rendering pipeline has been flushed)
Does the latter part mean once all animation frames are finished? You wouldn't want the callback invoked part way through the navigation animation.

**Comment by 0ttik** [comment]
Hi, @cohenadair!

You are right, in case of navigation animation it won't work. Although you can track it as well using `[ModalContext]` (obtaining it with `ModalContext.of(context)`) and then subscribing to animation state like this:

```
var modal = ModalContext.of(context);
modal.animation.addStatusListener(/*dont forget to unsubscribe in this callback*/);
```

And you just trigger map loading after this callback fires with value you need.

Docs:
1. https://api.flutter.dev/flutter/widgets/ModalRoute-class.html
2. https://api.flutter.dev/flutter/animation/Animation-class.html

**Comment by mellowcello77** [comment]
Finding the Google map too laggy to be useful in a list-view, will need to move it to a separate view for now. Above delays not working for me personally.

**Comment by niktob560** [comment]
@mellowcello77, did You tried `lite mode` for a list-view usage?

**Comment by mellowcello77** [comment]
@niktob560 I looked into it at the time, I think it was for Android only. I ended up using mapbox.

**Comment by ItsPierre** [comment]
@niktob560 I tried lite mode, but this did not solve the problem.

At the moment, the best workaround seems to be a FutureBuilder, that waits the 200 ms the page transition animation requires.
At this time I just show a shimmer-loading effect.
After this I initialize the GoogleMap.

Not perfect, but much better, than a lagging animation.

**Comment by darshankawar** [comment]
Verified this using a sample app on master version in debug mode and noticed the lag when map tries to open / initialize after directing through a route, as below:


https://user-images.githubusercontent.com/67046386/115868143-9db71c80-a459-11eb-8e1b-d8bd26a46d3a.mov


<details>
<summary> code sample </summary>

```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Google Maps Demo',
      home: MyHome()
    );
  }
}

class MyHome extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          onPressed:() {
            Navigator.push(
              context,
              MaterialPageRoute(builder: (context) => MapSample()),
            );
          },
          child: Text('Tap')
        )
      )
    );
  }
}

class MapSample extends StatefulWidget {
  @override
  State<MapSample> createState() => MapSampleState();
}

class MapSampleState extends State<MapSample> {
  Completer<GoogleMapController> _controller = Completer();

  static final CameraPosition _kGooglePlex = CameraPosition(
    target: LatLng(37.42796133580664, -122.085749655962),
    zoom: 14.4746,
  );

  static final CameraPosition _kLake = CameraPosition(
      bearing: 192.8334901395799,
      target: LatLng(37.43296265331129, -122.08832357078792),
      tilt: 59.440717697143555,
      zoom: 19.151926040649414);

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: GoogleMap(
        mapType: MapType.hybrid,
        initialCameraPosition: _kGooglePlex,
        onMapCreated: (GoogleMapController controller) {
          _controller.complete(controller);
        },
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: _goToTheLake,
        label: Text('To the lake!'),
        icon: Icon(Icons.directions_boat),
      ),
    );
  }

  Future<void> _goToTheLake() async {
    final GoogleMapController controller = await _controller.future;
    controller.animateCamera(CameraUpdate.newCameraPosition(_kLake));
  }
}


```

</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.2.0-11.0.pre.187, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.2.0-11.0.pre.187 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5526dcc24b (2 days ago), 2021-04-20 22:34:02 -0700
    • Engine revision 610fd039ae
    • Dart version 2.14.0 (build 2.14.0-18.0.dev)

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 90.0.4430.85

! Doctor found issues in 1 category.

```

</details>

**Comment by SergeyShustikov** [comment]
Any updates?

**Comment by martipello** [comment]
still seeing long wait times for google maps to load

**Comment by thealmamun** [comment]
still facing the issue..

**Comment by Adrek** [comment]
> Using this plugin really degrades the performance of the Widget tree for me... i had to put the map on a separate, isolate screen ;/

Please, could you share the code to generate an isolate screen?

**Comment by yacineblr** [comment]
Any news about this ?

**Comment by aleksandar-radivojevic** [comment]
Any updates

**Comment by kraa1055** [comment]
> I also have issues with Google map inside ListView, it's lagging always when I try to scroll, not just when initialising.

@aleksandar-radivojevic Same here. Do you have any luck with this issue ?

**Comment by filiph** [comment]
I was able to dig into this and here's a summary of my findings so far.

## Caveats

* I have only looked at Android so far. I haven't even tried to reproduce this issue on iOS — if you know it's there, too, please let me know.
* There were many red herrings along the way and I'm still not 100% sure I'm at the root cause, or that fixing the problem below won't leave some other problem persist (in fact, I do know that there are perf lags elsewhere, though they're an order of magnitude or two shorter than the one I'm focusing on)

## Methodology

* I created a [repro app](https://github.com/filiph/bug_maps_plugin_28493_perf_repro) that only has two screens: `MyHome` and `MapSample`. Clicking on a button in `MyHome` navigates to `MapSample`, and I made sure the transition animation is hard to miss (I use `CupertinoPageTransitionsBuilder()`, which slides the next page from the right).
* I instrumented the app, the Flutter engine, the Dart side of the maps plugin, and the Java side of the maps plugin.
* I looked at the DevTools Timeline and CPU profile (for the Dart parts) and I extracted timeline-like information from logs (for the instrumented Java parts).
* I made sure to measure only the first time the button is pressed. (The following transitions are much smoother.)

## Reproduced issue

On my Pixel 5, I can see up to 900ms jank after clicking the button in profile mode.

https://github.com/user-attachments/assets/3d45c048-de4d-4d22-b1e4-3a0e96261376

(The recording is not very readable, but note that the first time I tap, there's no animation — the only motion you see is after a while, and it's the Google Map being loaded tile by tile). The second time I tap, you see the transition.)

## Timeline

![Image](https://github.com/user-attachments/assets/ea2f5bbd-0d01-4c39-bfd7-5b62404a3bf5)

The first time the button is pressed, there are one or two Flutter frames (one of the heavy but not 900ms heavy), and then a long gap without any Dart code executed (the long "Frame Request Pending" / "Screen Display Lag" in the screenshot above).

I focus on the gap.

The last Dart timeline event before the gap is `sendCreateMessage start` and the first one after the gap is `sendCreateMessage end`, which I added around [this code in the Flutter engine](https://github.com/flutter/flutter/blob/be698c48a6750c8cb8e61c740ca9991bb947aba2/packages/flutter/lib/src/services/platform_views.dart#L1047-L1056) (in `SurfaceAndroidViewController`). And that thing simply invokes a method through a platform channel. So, in my interpretation at least, the gap doesn't have anything to do with any Dart code.

## Platform-side

When the Java side receives the create message, it eventually runs [`GoogleMapController.onCreate()`](https://github.com/flutter/packages/blob/ab44c26de09834d0c2fc0716271f48c86ce1e560/packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java#L457-L463) (through `lifecycle.addObserver(this)` in `GoogleMapController.init()`).

And that code runs [`MapView.onCreate()`](https://developers.google.com/android/reference/com/google/android/gms/maps/MapView#onCreate(android.os.Bundle)), the culprit. This is where the device spends about 900ms hogging the thread — at least as far as I understand. 

Google Maps isn't open source, but we can look at the decompiled source to see what's going on in `MapView.onCreate()`:

```java
    public void onCreate(@Nullable Bundle savedInstanceState) {
        StrictMode.ThreadPolicy var2 = StrictMode.getThreadPolicy();
        StrictMode.setThreadPolicy((new StrictMode.ThreadPolicy.Builder(var2)).permitAll().build());

        label71: {
            Throwable var10000;
            label75: {
                try {
                    this.zza.onCreate(savedInstanceState);
                    var9 = this.zza.getDelegate();
                } catch (Throwable var8) {
                    var10000 = var8;
                    boolean var10001 = false;
                    break label75;
                }

                if (var9 != null) {
                    break label71;
                }

                label66:
                try {
                    DeferredLifecycleHelper.showGooglePlayUnavailableMessage(this);
                    break label71;
                } catch (Throwable var7) {
                    var10000 = var7;
                    boolean var10 = false;
                    break label66;
                }
            }

            StrictMode.setThreadPolicy(var2);
            throw var10000;
        }

        StrictMode.setThreadPolicy(var2);
    }
```

To me, this looks like the method saves the current thread policy to `var2`, then changes the thread policy through `setThreadPolicy()`, then does its thing, and then returns the thread policy to its original value (`var2`).

By applying [`permitAll()`](https://developer.android.com/reference/android/os/StrictMode.ThreadPolicy.Builder#permitAll()) on the thread policy, the code "disables detection of everything".

From the [docs on StrictPolicy](https://developer.android.com/reference/android/os/StrictMode):

> StrictMode is most commonly used to catch accidental disk or network access on the application's main thread, where UI operations are received and animations take place.

This, to me, suggests that the contents of `MapView.onCreate()` do something that would otherwise create a warning. So the code above suppresses it.

And so I think that something in `MapView.onCreate` simply hogs the main thread.

## Solutions explored

I thought maybe simply deferring `MapView.onCreate` would help. Maybe we can return from `sendCreateMessage()` sooner (before the MapView is created). That way, the Flutter app would continue running. But that doesn't really help. As soon as `MapView.onCreate` is called, it hogs the thread regardless of when it's running, preventing the Flutter app from animating. So it only defers the problem for a bit later.

`MapView.onCreate` must be called on the main thread, so there's no way to parallelize it.

[One StackOverflow answer](https://stackoverflow.com/a/27599390) mentioned that this simply happens in debug mode and/or that it happens when the development device is plugged into the USB. But I can reproduce in release & profile modes, disconnected.

## Why so silent on the Android side?

Question: Is this happening in pure Android apps, too? If not, what is it about the Flutter engine + plugin mechanism that makes this a problem? I did find an [Android issue about this](https://issuetracker.google.com/issues/80147879), but it's surprisingly old (2018) and has surprisingly few upvotes (33). It was also marked as _Won't Fix (Obsolete)_ in 2023.

The one workaround on the Android side that I could find is in the [aforementioned StackOverflow question](https://stackoverflow.com/questions/26265526/how-to-improve-fragment-loading-speed/29246677#29246677) (linked from the aforementioned Android issue). It calls for "warming up" MapView create by calling it at the app's startup. Something like:

```java
    // Fixing Later Map loading Delay
    new Thread(new Runnable() {
        @Override
        public void run() {
            try {
                MapView mv = new MapView(getApplicationContext());
                mv.onCreate(null);
                mv.onPause();
                mv.onDestroy();
            }catch (Exception ignored){

            }
        }
    }).start();
```

Question: do Android apps out there in the wild actually do this?


## Solutions to be explored
 
I don't think there's anything that can be done on the Flutter engine or, more generally, the Dart side of things, to really solve the issue — but there are a few ways to at least partly address the pain point.

1. Change `google_maps_flutter_android` to use the **warm up hack at app startup** (e.g. during plugin registration).
2. Let developers run the **warm up on demand**. For example, add a static `GoogleMap.warmUp()` method.
3. Automatically detect that a `GoogleMap` is being intialized for the first time (in the app's lifetime) and **defer initialization** until after animation is finished. In the meantime, simply don't draw the map, or even let users specify a placeholder widget.
4. **Simply document** the issue and list workarounds (e.g. modifying `MainActivity` to run the warmup hack on app startup, or only adding `GoogleMap` to the widget tree when animations are finished).

I'd like to understand what the maintainers and the community think about these solutions, or if I missed some better alternatives.

**Comment by JRR-OSU** [comment]
Thanks for the deep dive! @filiph 

Since the original days of this issue some 6 years ago, a lot has changed in the world, except Google Maps' performance 😅. In the past few years since commenting, I have been doing a lot of native Android development with Kotlin / Compose professionally, and can say definitively this type of workaround is absolutely used in production. Thread management on Android has improved since the Java days with the advent of Coroutines. We use those to launch asynchronous lower priority invocation and warmup of our Map composables' internals (still on the main thread), but without causing massive jank via side-effects in Compose. This is how we work around it there. 

We also have used the Compose library for Maps from Google. That has initialization slowdown too though not nearly to the degree seen in Flutter apps. It seems like when the AndroidView wrapping the MapView calls onCreate, that all is happening asynchronously. 

https://github.com/googlemaps/android-maps-compose/blob/main/maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

I'd personally recommend some kind of warmup method that can do the initialization without rendering anything. Bonus points if the warmup can happen asynchronously too!

**Comment by filiph** [comment]
Thanks for the info, @JRR-OSU! 

@stuartmorgan-g, it looks like a good way to address this issue would be to give Flutter developers a way to "warm up" Google Maps on demand. Is this something you'd consider merging if I made a PR?

Here are the PROs and CONs I can see from where I'm standing.

PROs:

1. Addresses one of the most-voted-for and infamous issues on this plugin — i.e. clearly something people struggle with.
2. Follows what native Android developers are doing to address the same issue.
3. Doesn't require a whole lot of new code nor changes to the engine; can be done today.

CONs:

1. "Gives up" on fixing the underlying issue (Google Maps hogging the main thread) - instead gives people a workaround
2. It could be argued that Flutter developers already have the ability to warm up (by changing android .java or .kt files in their Flutter project)

**Comment by stuartmorgan-g** [comment]
> 1. Addresses one of the most-voted-for and infamous issues on this plugin — i.e. clearly something people struggle with.

Maybe. I would imagine that most people are voting to express that they want there not to be lag, not to express that they want the same amount of lag at a different time.

That's not to say that it isn't potentially useful, but it seems unlikely that it will actually result in most people who voted for it feeling that it's addressed.
  
> 1. "Gives up" on fixing the underlying issue (Google Maps hogging the main thread) - instead gives people a workaround

I don't follow. How does providing a Dart affordance for an existing workaround impact future Google Maps SDK development? If this behavior is inherent to the SDK, nobody on the Flutter team or in the Flutter community can change that.

I'd like to understand the problem more before we add a new API though. The fact that the links you've found are all old, and the issue is marked obsolete, makes it unclear to me to what extent this should still be an issue. The Android SDK has an initialization system now; did your sample app call it? Did that change how much work onCreate does?

**Comment by filiph** [comment]
> Maybe. I would imagine that most people are voting to express that they want there not to be lag, not to express that they want the same amount of lag at a different time.

Help me understand. Are you saying that developers who voted on this issue aren't interested in ways to mitigate the lag unless it completely removes the lag? Why do you think so?

We have a chance to give developers the option to warm up a subsystem so that it doesn't lag at one of the worst times an app can lag (just after user interaction; when the app is most likely transitioning to a new page; when an animation is most likely running). See also [time scales in UX](https://www.nngroup.com/articles/powers-of-10-time-scales-in-ux/) by Jakob Nielsen.

From this perspective, I think it's a bit strange to dismiss the mitigation with "they want the same amount of lag at a different time". I obviously can't talk for anyone else but FWIW, when I can't avoid lag in an app, I do try to manage when it happens. During transition, after human interaction, is much worse than during app startup, for example.

Yes, it would be ideal to have no lag at all — we can all agree. But that's not clearly possible even when using native Android (see https://github.com/flutter/flutter/issues/28493#issuecomment-2920703870 above).

> That's not to say that it isn't potentially useful, but it seems unlikely that it will actually result in most people who voted for it feeling that it's addressed.

Are you merely expressing doubt that this mitigation will be enough for the people who voted for this issue? Or are you saying I should create a separate issue only for people who want to mitigate the issue, and keep this one for the developers that you assume are only waiting for solution to the underlying problem?

> > 1. "Gives up" on fixing the underlying issue (Google Maps hogging the main thread) - instead gives people a workaround
> 
> I don't follow. How does providing a Dart affordance for an existing workaround impact future Google Maps SDK development? If this behavior is inherent to the SDK, nobody on the Flutter team or in the Flutter community can change that.

Sorry for not being clear. I was trying to express that by going with the mitigation, I would "give up" on trying to find a fix for the underlying issue. I was not talking about others. 

> I'd like to understand the problem more before we add a new API though. The fact that the links you've found are all old, and the issue is marked obsolete, makes it unclear to me to what extent this should still be an issue. The Android SDK has an initialization system now; did your sample app call it? Did that change how much work onCreate does?

Thanks for mentioning the initialization system ([`MapsInitializer`](https://developers.google.com/android/reference/com/google/android/gms/maps/MapsInitializer?hl=en#initialize(android.content.Context,%20com.google.android.gms.maps.MapsInitializer.Renderer,%20com.google.android.gms.maps.OnMapsSdkInitializedCallback))). It's already being used by the package, and in fact the `example` in `google_maps_flutter_android` calls it from `main()` ([here](https://github.com/flutter/packages/blob/6d3aaf4b552374d74ad87a2c3935c6ed398ffc70/packages/google_maps_flutter/google_maps_flutter_android/example/lib/main.dart#L86-L110)). Maybe I'm holding it wrong, but in my testing, adding the initialization step doesn't have any significant effect on the lag.

Without pre-initialization:

![Image](https://github.com/user-attachments/assets/f586a422-7530-4bae-ac5c-690b0559e95c)

With pre-initialization:

![Image](https://github.com/user-attachments/assets/53b081b6-8cd9-4901-879f-c8b9a317fc59)

I was worried that maybe I'm not initializing the latest renderer (which is supposed to be much more performant than the old one) but I just confirmed I get `AndroidMapRenderer.latest` from `initializeWithRenderer()`.

**Comment by stuartmorgan-g** [comment]
Since it's still happening with the latest renderer and having already called the initializer, I'm fine with adding an API that would allow pre-creating a temporary invisible map.

**Comment by filiph** [comment]
Ok, I have a question or two before I create the PR, and some results that I'd like to show so that you can decide if the improvement is worth it.

Here's the improvement in video:

https://github.com/user-attachments/assets/3491f78f-06d9-4be8-b39b-1a49b9ef2c0c

Note:

* The circular progress indicator is there just as a way to visualize jank. Nothing is actually loading — it's just the easiest way I know to show continuous animation.
* I sabotaged the demo by missing the tap button the first time, and I noticed too late. Now I don't want to retake the video just to remedy this. Hope you can abstract from that issue.

Actually, it's much easier to see all the subtle differences when the video is 4x slower:

https://github.com/user-attachments/assets/7b7c1c0c-9bf2-4aa0-9e7b-f83d3f979ca8

The most important issue being addressed here is the complete lack of any page transition on the left. The warmed-up version also janks, but not nearly as much.

Finally, here's the effect of the warmup phase on app startup:

https://github.com/user-attachments/assets/498b935b-0a3b-46d7-a88d-886150537018

As @stuartmorgan-g says, this is not actually removing the jank, it's just moving it to elsewhere. I do think it's better to give the developer an option of when and where to have the largest jank. And I think that for most apps, it's best at the very beginning (as long as we're still talking about sub-second janks, of course, and not some long loading screen).

Now, an API design question. I originally thought that I'd introduce some kind of separate `warmup()` function. But it's not clear where to put it, and I think it _maybe_ makes more sense to bundle it with the existing `initializeWithRenderer()` method. What do you think, @stuartmorgan-g?

```dart
  /// Requests Google Map Renderer with [AndroidMapRenderer] type.
  ///
  /// See https://pub.dev/packages/google_maps_flutter_android#map-renderer
  /// for more information.
  ///
  /// The renderer must be requested before creating GoogleMap instances as the
  /// renderer can be initialized only once per application context.
  /// Throws a [PlatformException] if method is called multiple times.
  ///
  /// The returned [Future] completes after renderer has been initialized.
  /// Initialized [AndroidMapRenderer] type is returned.
  ///
  /// When [warmup] is `true`, the plugin will also warm up the Android
  /// Google Maps SDK just after initializing it. This is so that
  /// the first showing of a [GoogleMap] widget doesn't introduce too much jank.
  /// This is a trade-off, since the jank isn't eliminated, only moved.
  Future<AndroidMapRenderer> initializeWithRenderer(
      AndroidMapRenderer? rendererType,
      {bool warmup = false}) async {
    // ...
  }
```

The advantages of bundling `warmup` in `initializeWithRenderer`:

1. Minimal change to API - just a new optional argument
2. No way for the developer to screw this up (e.g. calling warmup before initialize)
3. Slightly easier to implement
4. Warming up after initialization is what I expect 99% of developers to want to do

The disadvantages:

1. Less control - the warmup is called just after initialization, but that initialization is asynchronous, and could therefore happen at a bad time. Some developers might want to wait for a more opportune time.

**Comment by stuartmorgan-g** [comment]
> 4\. Warming up after initialization is what I expect 99% of developers to want to do

I don't see any reason to expect that developers with apps that only show maps in specific flows wouldn't want to separate renderer initialization (presumably cheap, but async with a server component) with the unrelated warmup step (slow/janky).

**Comment by filiph** [comment]
> > 4. Warming up after initialization is what I expect 99% of developers to want to do
> 
> I don't see any reason to expect that developers with apps that only show maps in specific flows wouldn't want to separate renderer initialization (presumably cheap, but async with a server component) with the unrelated warmup step (slow/janky).

Ok, I see a reason or two but moving on. So you want me to create a new method on the same object, then? Any preference on naming? Should it throw if used before initializeWithRenderer or should it call that method automatically when appropriate?

**Comment by stuartmorgan-g** [comment]
> So you want me to create a new method on the same object, then? Any preference on naming?

It would be a method on `GoogleMapsFlutterAndroid`, yes. I'll defer to the Android subteam on naming; that can be figured out in review as it's relatively minor.

> Should it throw if used before initializeWithRenderer or should it call that method automatically when appropriate?

I don't understand the question; `initializeWithRenderer` isn't a required method, so I wouldn't expect it to do either of those things.
